torch/_inductor/codegen/cpp.py

```diff
@@ -3863,25 +4031,120 @@ def wrap_fn(fn):
+                    if_prefix = "C10_UNLIKELY"
+                    stack.enter_context(code.indent())
+                    body = kernel.gen_body()
+                    if kernel.inner_itervars:
+                        for idx in kernel.inner_itervars:
+                            start, end = kernel.active_ranges[idx]
+                            body = transform_kernel_codes_under_inner_loop(
+                                body, idx, f"{idx}_tail", start, end
+                            )
+                    code.splice(body)
+
+    def aggregate_reduction_buffers(
+        self, inner_loop_reduction_outer_not: bool, outer_loop: Optional["LoopLevel"]
+    ):
+        def get_buffer_from_main_loop_kernel(attr: str):
+            main_kernel = self.kernels[0]
+            if attr == "reduction_prefix":
+                main_kernel.finalize_reduction_prefix()
+            getattr(self, attr).splice(getattr(main_kernel, attr))
+
+        def aggregate_reduction_prefix_suffix(outer_loop: "LoopLevel"):
+            assert len(self.kernels) >= 2
+            main_loop_kernel = self.kernels[0]
+            tail_loop_kernel = self.kernels[-1]
+            assert isinstance(main_loop_kernel, CppVecKernel)
+
+            # Prefix
+            if type(tail_loop_kernel) == CppKernel:
+                # if tail loop kernel is a scalar kernel, we need to extend tmp_acc -> tmp_acc_arr[] t
+                # hold the temporal inner loop acc result for outer tail loop
+                tail_loop_kernel.finalize_reduction_prefix(
+                    main_loop_kernel.tiling_factor
+                )
+                main_loop_kernel.finalize_reduction_prefix()
+                self.reduction_prefix.splice(
+                    tail_loop_kernel.reduction_prefix
+                    + main_loop_kernel.reduction_prefix
+                )
+            else:
+                get_buffer_from_main_loop_kernel("reduction_prefix")
+
+            # Suffix
+            suffix_buf = BracesBuffer()
+            with contextlib.ExitStack() as stack:
+                if main_loop_kernel.codegen_conditions(
+                    suffix_buf, "C10_LIKELY", outer_loop.var
+                ):
+                    stack.enter_context(suffix_buf.indent())
+                    suffix_buf.splice(main_loop_kernel.reduction_suffix)
+            with contextlib.ExitStack() as stack:
+                if tail_loop_kernel.codegen_conditions(
+                    suffix_buf, "C10_UNLIKELY", outer_loop.var
+                ):
+                    stack.enter_context(suffix_buf.indent())
+                    if type(tail_loop_kernel) == CppKernel:
+                        reduction_vars = tail_loop_kernel.reduction_var_names
+                        for name in reduction_vars:
+                            new_name = f"{name}_arr[{outer_loop.var}_tail - {cexpr_index(outer_loop.tiling_offset)}]"
+                            replace_acc_name(tail_loop_kernel.stores, name, new_name)
```

Comment:
Do we still need the logic of replacing a generated buffer after introducing the design of lazy generation with callbacks?

Reply:
"replace_acc_name" handles simple cases which just replace an acc's name for store buffers and reduction suffix buffers.
```
tmp0 += tmp_acc0; -> tmp0 += tmp_acco_arr[x];
```
" lazy generation with callbacks" handles more complex case for `reduction_prefix`. It need also extend the initialization
```
float tmp_acc0 = 0; ->
float tmp_acc0_arr;
for i in range(size):
    tmp_acc0_arr[i] = 0;
  
```

Reply:
But would that make the logic simpler if we also leverage the new lazy mechanism to handle the `replace_acc_name` logic?

Reply:
Hi, @jgong5. I feel it is simpler to use `replace_acc_name`, since 
(1) This functionality is pretty simple, just replace the acc var names
(2) This function can be reused by `store` buffer and `reduction_suffix` buffer. If we apply lazy mechanism, we will need to introduce `store_generator` and `reduction_suffix_generator`.
I think it might be better to leverage lazy mechanism when we need more complex case to modify `store` and `reduction_suffix`.